nucypher/characters.py

```diff
@@ -558,12 +586,12 @@ def follow_treasure_map(self, hrac, using_dht=False):
                         number_of_known_treasure_ursulas += 1
                 newly_discovered_nodes.update(new_nodes)
 
-            self.known_nodes.update(newly_discovered_nodes)
+            self._known_nodes.update(newly_discovered_nodes)
             return newly_discovered_nodes, number_of_known_treasure_ursulas
         else:
             for ursula_interface_id in self.treasure_maps[hrac]:
                 pubkey = UmbralPublicKey.from_bytes(ursula_interface_id)
-                if ursula_interface_id in self.known_nodes:
+                if ursula_interface_id in self._known_nodes:
```

Comment:
Since blockchain sampling returns ethereum addresses, We need a way to quickly lookup known nodes by their addresses, while still providing compatibility for a federated version where known nodes is keyed differently.

Reply:
So we do mix both types, right?

Reply:
Yeah, we've gone around and around about this, but I think the current solution is ideal.  We talked about having two dicts (one for each), but now instead we provide a way (`public_address`) to get an ethereum-address like construct from an actor who doesn't have a blockchain presence, using their `UmbralPublicKey`.

Reply:
@cygnusv: No, we don't mix both types - we use a 20 byte "address-style" representation in either case.